File: AlgorithmDataStructure/LeetCode/83-DeleteDuplicates/delete.c

#include <stdio.h>



struct ListNode {
    int val;
    struct ListNode *next;
};
void dump(struct ListNode *head) {
    if (head == NULL) { printf("list null! \n"); }
    struct ListNode *node = head;
    while (node != NULL) {
        printf("%d -> ", node->val);
        node = node -> next;
    }
    printf("\n");
}

struct ListNode* deleteDuplicates(struct ListNode* head){
    if (head == NULL) { return head; }
    
    struct ListNode *temp = head;
    while (temp != NULL && temp -> next != NULL) {
        if (temp->val == temp ->next->val) {
            struct ListNode *node = temp->next;
            temp->next = node->next;
            node->next = NULL;
            //free(node);
        } else {
            temp = temp->next;
        }
    }
    
    return head;
}

int main(int argc, const char * argv[]) {
    
    struct ListNode l1 = {3, NULL};
    struct ListNode l2 = {3, &l1};
    struct ListNode l3 = {2, &l2};
    struct ListNode l4 = {2, &l3};
    struct ListNode l5 = {2, &l4};
    struct ListNode l6 = {1, &l5};
    dump(&l6);
    
    struct ListNode *l = deleteDuplicates(&l6);
    dump(l);
    
    return 0;
}

File: AlgorithmDataStructure/LeetCode/83-DeleteDuplicates/delete.swift

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
        
    }
    
    func dump() {
        var l: ListNode? = self
        while l != nil {
            print("\(l!.val)")
            l = l!.next
        }
    }
}



class Solution {
    func deleteDuplicates(_ head: ListNode?) -> ListNode? {
        var temp = head
        while temp != nil && temp!.next != nil {
            if temp!.val == temp!.next!.val {
                let node = temp!.next
                temp!.next = node?.next
                node!.next = nil
            } else {
                temp = temp!.next
            }
        }
        return head
    }
}


let s = Solution()

var l0 = ListNode(3)
var l1 = ListNode(3)
l1.next = l0
var l2 = ListNode(2)
l2.next = l1
var l3 = ListNode(2)
l3.next = l2
var l4 = ListNode(1)
l4.next = l3

l4.dump()
print("-----")
var r = s.deleteDuplicates(l4)
r!.dump()
